features/filesystem/bd/BlockDevice.h

```diff
@@ -217,6 +218,10 @@ class BlockDevice
             size % get_erase_size() == 0 &&
             addr + size <= this->size());
     }
+
+protected:
+    uint32_t _init_ref_count;
```

Comment:
Sorry I didn't see this earlier. We should not have state in interfaces classes.

This adds a RAM cost to every block device class. Even ones with their own form of synchronization (ie SD, SPIF, others outside of tree).

At worst this can discourage users from applying the interface to their own code. Interfaces work best when a class satisfies as many interfaces as possible, allowing code reuse for functions that use that interface. If there's extra state in an interface, users may not adopt the interface because of the code cost. Even if it makes sense otherwise.

This is what happened with the FileLike/FileSystemLike classes. State was introduced for constructing the file tree, which on paper was a good idea, but now the interface is avoided and the poorer-named-but-stateless FileHandle class is used instead. (We still have to put up with this because of backwards compatibility, but it was a lesson learned.)

---

I believe this can be moved into each of the implementing classes without any cost?

Reply:
Will move. Certainly now that not all classes implement it.

Reply:
Thanks. The FileLike/FileSystemLike classes have been frustrating, and I didn't want to see us repeat it.